### PR TITLE
Bug/#25 write std out err to log file

### PR DESCRIPTION
**Pull request by tomuben**
#25 : Redirected build logs to $job_dir/exaslct.log
    
    Printing logs to the console only has the disadvantages
    of analyzing possible issues reported from 3rd party.
    Having logs persistently gives the possibility to share the logs
    later among  different teams.
    For now, only the build logs are redirected, however,
    in the future any exaslct command can be easily redirecty,
    by encapsulating the creation function of the task with
    logging_redirection.create_docker_build_creator()

### CODE REVIEW

**Comment by tkilias** [comment]
Does the tee still duplicates outputs?

**Comment by tomuben** [comment]
> @tomuben Does the tee still duplicates outputs?

No, it was a mistake on my side, when printed to both stdout and stderr.

File: exasol_script_languages_container_tool/cli/commands/build.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import sys
```
Comment (tkilias) [inline exasol_script_languages_container_tool/cli/commands/build.py:1]: Do we still need sys here?

File: exasol_script_languages_container_tool/lib/utils/logging_redirection.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    def write(self, data):
+        self.file.write(data)
+        self.stdout.write(data)
+
+    def flush(self):
+        self.file.flush()
+
+
+def create_docker_build_creator(task_creator):
```
Comment (tkilias) [inline exasol_script_languages_container_tool/lib/utils/logging_redirection.py:29]: hm, I think, the name doesn't really fit. Maybe something like that `LogRedirectorTaskCreatorWrapper`

File: exasol_script_languages_container_tool/cli/commands/build.py [hunk 3]
```diff
@@ -70,7 +72,9 @@ def build(flavor_path: Tuple[str, ...],
     set_docker_repository_config(target_docker_password, target_docker_repository_name, target_docker_username,
                                  target_docker_tag_prefix, "target")
     set_job_id(DockerBuild.__name__)
-    task_creator = lambda: DockerBuild(flavor_paths=list(flavor_path), goals=list(goal), shortcut_build=shortcut_build)
+    task_creator = create_docker_build_creator(lambda: DockerBuild(flavor_paths=list(flavor_path),
+                                                                   goals=list(goal), shortcut_build=shortcut_build))
+
     success, task = run_task(task_creator, workers, task_dependencies_dot_file)
```
Comment (tkilias) [inline exasol_script_languages_container_tool/cli/commands/build.py:78]: We should print the log file location after `run_task` ran.

File: exasol_script_languages_container_tool/lib/utils/logging_redirection.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    # Provides redirection of ALL logs to exaslct.log in the tasks job directory
+
+    # Parameters:
+    #   task_creator (function): creator function of the task
+    def create_docker_build():
+        task = task_creator()
+        tee = _Tee(f'{task.get_log_path()}/exaslct.log', "w")
```
Comment (tkilias) [inline exasol_script_languages_container_tool/lib/utils/logging_redirection.py:36]: Do you have an example for the log file path. `task.get_log_path()` might be a subdirectory of the job-directory.

**Review by tkilias — CHANGES_REQUESTED** [review]

File: exasol_script_languages_container_tool/cli/commands/build.py [hunk 5]
```diff
@@ -70,7 +73,11 @@ def build(flavor_path: Tuple[str, ...],
...
+    print(f'Build log can be found at:{get_log_path(task)}')
```
Comment (tkilias) [inline exasol_script_languages_container_tool/cli/commands/build.py:80]: I think, we need this also for the other commands

File: exasol_script_languages_container_tool/lib/utils/logging_redirection.py (hunk 4 above)
Comment (tomuben) [reply exasol_script_languages_container_tool/lib/utils/logging_redirection.py]: Here is one
 .build_output/jobs/2021_05_31_11_01_19_DockerBuild/outputs/DockerBuild_1c5a98ee16/logs/exaslct.log

File: exasol_script_languages_container_tool/lib/utils/logging_redirection.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    def flush(self):
+        self.file.flush()
+
+
+def log_redirector_task_creator_wrapper(task_creator):
+    # Provides redirection of ALL logs to exaslct.log in the tasks job directory
+
+    # Parameters:
+    #   task_creator (function): creator function of the task
+    def create_docker_build():
```
Comment (tkilias) [inline exasol_script_languages_container_tool/lib/utils/logging_redirection.py:36]: Now, this one also needs a better name.

File: exasol_script_languages_container_tool/lib/utils/logging_redirection.py (hunk 4 above)
Comment (tkilias) [reply exasol_script_languages_container_tool/lib/utils/logging_redirection.py]: thinking about, if we should show the whole path to exaslct or only the following part

**Review by tkilias — CHANGES_REQUESTED** [review]
Good, most things are done.

File: exasol_script_languages_container_tool/lib/utils/logging_redirection.py (hunk 4 above)
Comment (tomuben) [reply exasol_script_languages_container_tool/lib/utils/logging_redirection.py]: Ok, will keep as is then.

File: exasol_script_languages_container_tool/lib/utils/logging_redirection.py (hunk 6 above)
Comment (tomuben) [reply exasol_script_languages_container_tool/lib/utils/logging_redirection.py]: done

**Review by tkilias — APPROVED** [review]

**Pull request merged**